File: Algo Java/binarysearchtrees/BinarySearchTree.java
/**
 *
 * @author Gaurab R. Gautam
 */
package binarysearchtrees;

public class BinarySearchTree<T extends Comparable<? super T>> 
{
    Node<T> root = null;
    private int size = 0;
    private int height = 0;
    
    public BinarySearchTree(T[] vals)
    {
        buildTree(vals);
    }
    
    public BinarySearchTree()
    {
        
    }
    
    public Node<T> getroot() {
        return this.root;
    }
    
    public void setRoot(Node<T> val) {
        this.root = val;
    }
    
    public int getSize() {
        return this.size;
    }
    
    public void setSize(int val) 
    {
        this.size = val;
    }
    
    // upper bound of height of red black tree
    public int getHeight() {
        this.height = 2 * (int)Math.floor(Math.log(this.size + 1)/Math.log(2));
        return this.height;
    }
    
    public boolean isEmpty() 
    {
        return this.size == 0;
    }
    
    public Node<T> getRoot()
    {
        return this.root;
    }
    
    private void buildTree(T[] vals)
    {
        for (int i = 0; i < vals.length; i++)
        {
            insert(new Node<>(vals[i]));
        }
    }
    
    public void inorderTreeWalk(Node<T> node)
    {
        if (node != null)
        {
            inorderTreeWalk(node.left);
            System.out.print(node.key + " ");
            inorderTreeWalk(node.right);
        }
    }
    
    public Node<T> search(Node<T> node, T key)
    {
        if (node == null || node.key.equals(key))
        {
            return node;
        }
        else if (key.compareTo(node.key) < 0)
        {
            return search(node.left, key);
        }
        else
        {
            return search(node.right, key);
        }
        
    }
    
    public Node<T> minimum(Node<T> node)
    {
        while (node.left != null)
        {
            node = node.left;
        }
        
        return node;
    }
    
    public Node<T> maximum(Node<T> node)
    {
        while (node.right != null)
        {
            node = node.right;
        }
        
        return node;
    }
    
    public Node<T> successor(Node<T> node)
    {
        if (node.hasRight())
        {
            return minimum(node.right);
        }
        else
        {
            Node<T> parent = node.parent;
            
            while (parent != null && node.isRightChild())
            {
                node = parent;
                parent = node.parent;
            }
            
            return parent;
        }
    }
    
    public Node<T> predecessor(Node<T> node)
    {
        if (node.hasLeft())
        {
            return maximum(node.left);
        }
        else
        {
            Node<T> parent = node.parent;
            
            while (parent != null && node.isLeftChild())
            {
                node = parent;
                parent = node.parent;
            }
            
            return parent;
        }
    }
    
    private void transplant (Node<T> u, Node<T> v)
    {
        if (u.parent == null)
        {
            root = v;
        }
        else if (u.isLeftChild())
        {
            u.parent.left = v;
        }
        else
        {
            u.parent.right = v;
        }
        
        if (v != null)
        {
            v.parent = u.parent;
        }
    }
    
    public boolean delete(T key)
    {
        Node<T> node = get(key);
        
        if (node == null)
        {
            System.out.println(key + " not found!");
            return false;
        }
        
        return delete(node);
    }
    
    public boolean delete(Node<T> node)
    {
        boolean deleted = false;
        
        if (!node.hasLeft())
        {
            transplant(node, node.right);
            node = null;
            this.size -= 1;
            
            deleted = true;
        }
        else if (!node.hasRight())
        {
            transplant(node, node.left);
            node = null;
            this.size -= 1;
            deleted = true;
        }
        else if (node.hasLeft() && node.hasRight())
        {
            Node<T> successor = minimum(node.right);
            
            if (successor.parent != node)
            {
                transplant(successor, successor.right);
                successor.right = node.right;
                successor.right.parent = successor;
            }
            
            transplant(node, successor);
            successor.left = node.left;
            node.left.parent = successor;
            
            node = null;
            this.size -= 1;
            
            deleted = true;
        }
        
        return deleted;
    }
    
    public Node<T> insert(T val)
    {
        Node<T> newNode = new Node(val);
        insert(newNode);
        
        return newNode;
    }
    
    private void insert(Node<T> newNode)
    {
        Node<T> parent = null;
        Node<T> node = root;
        
        while (node != null)
        { 
            parent = node;
            
            if (newNode.key.compareTo(node.key) < 0)
            {
                node = node.left;
            }
            else
            {
                node = node.right;
            }
        }
        
        newNode.parent = parent;
        
        if (parent == null)
        {
            root = newNode;
        }
        else if (newNode.key.compareTo(parent.key) < 0)
        {
            parent.left = newNode;
        }
        else
        {
            parent.right = newNode;
        }
        
        this.size += 1;
    }
    
    public int getLevel(Node<T>node) {
        if (this.root == null || node == null) {
            return -1;
        }
        
        Node<T> iter = this.root;
        int level = 0;
        
        while (iter != null && iter != node) {
            if (node.key.compareTo(iter.key) < 0) {
                iter = iter.left;
                level += 1;
            }
            else {
                iter = iter.right;
                level += 1;
            }
        }
        
        return level;
    }
    
    public Node<T> get(T val) {
        return this.search(this.root, val);
    }
    
    public boolean has(T val) {
        return (this.search(this.root, val) != null);
    }
    
}

File: Algo JS/binarysearchtree.js
/************************************************************************
 * BinarySearchTree class
 * @param {Array} A - input array
 ************************************************************************/

// import modules
const BinaryTree = require('./binarytree');

module.exports = class BinarySearchTree extends BinaryTree {
    constructor() {
        super();
    }
    
    build(array) {
        array.forEach( element => {
            this.insert(element);
        });
    }
    
    insert(val) {
        let node = this.root;
        let parent = null;
        let newNode = new BinaryTree.Node(val);
            
            
        // empty tree
        if (node === null) {
            this.root = newNode;
            //this.root.level = 0;
        }
        else {
            while (node !== null) {
                parent = node;
                    
                if (val < node.key) {
                    node = node.left;
                }
                else {
                    node = node.right;
                }
            }
            
            if (val < parent.key) {
                parent.left = newNode;
            }
            else {
                parent.right = newNode;
            }
            
            newNode.parent = parent;
            //newNode.level = parent.level + 1;
        }
        
        this.size += 1;
        
        return newNode;
    }
    
    delete(item) {
        let node = this.get(item);
        
        if (node === null) {
            return (item + " not found!");
        }
        
        this.deleteNode(node);
    }
    
    deleteNode(node) {
        // case 1: node.left = nil
        if (node.left === null) {
            this.transplant(node, node.right);
            //node.right.level = node.level;
            
            return (node.key + " deleted!");
        }
        
        // case 2: node.right = nil
        else if (node.right === null) {
            this.transplant(node, node.left);
            //node.left.level = node.level;
                        
            return (node.key + " deleted!");
        }
        else {
            let successor = this.successor(node);
            
            // case 3: node.right = successor of node
            if (successor.parent === node) {
                this.transplant(node, successor);
                successor.left = node.left;
                successor.left.parent = successor;
                //successor.level = node.level;
            }
            
            // case 4: node's successor is in right subtree of node.right
            // successor's left child is nill, so consider right child only
            else if (successor.parent !== node) {
                this.transplant(successor, successor.right);
                successor.right = node.right;
                successor.right.parent = successor;
                //successor.right.level = successor.level;
                
                this.transplant(node, successor);
                successor.left = node.left;
                successor.left.parent = successor;
                //successor.level = node.level;
            }
            
            this.size -= 1;
            
            return (node.key + " deleted!");
        }
    }
    
    // rearrange parent pointer of deleted node-node2
    transplant(node1, node2) {
        if (node1.hasParent()) {
            this.root = node2;
        }
        else if (node1.isLeftChild()) {
            node1.parent.left = node2;
        }
        else {
            node1.parent.right = node2;
        }
        
        if (node2 !== null) {
            node2.parent = node1.parent;
        }
    }
    
    treeMinimum(node) {
        while (node.left !== null) {
            node = node.left;
        }
        
        return node;
    }
    
    treeMaximum(node) {
        while (node.right !== null) {
            node = node.right;
        }
        
        return node;
    }
    
    successor(node) {
        if (node.hasRight()) {
            return this.treeMinimum(node.right);
        }
        
        while (node.hasParent()) {
            node = node.parent;
            
            if (node.hasParent() && node === node.parent.right) {
                return node;
            }
        }
        
        return null;
    }
    
    predecessor(node) {
        if (node.hasLeft()) {
            return this.treeMaximum(node.left);
        }
        else {
            let parent = node.parent;
            
            while (parent !== null && node.isLeftChild()) {
                node = parent;
                parent = node.parent;
            }
            
            return parent;
        }
    }
    
    search(node, val) {
        if (node === null || (node.key === val)) {
            return node;
        }
        
        if (val < node.key) {
            return this.search(node.left, val);
        }
        else {
            return this.search(node.right, val);
        }
    }
    
    inOrderTreeWalk(node) {
        if (node === null) {
            return;
        }
        
        this.inOrderTreeWalk(node.left);
        console.log(node.key + ", level = " + this.getLevel(node) + ", left: " +
                (!node.hasLeft() ? "\u03A6" : node.left.key) + ", right: " +
                (!node.hasRight() ? "\u03A6" : node.right.key) + ", parent: " +
                (!node.hasParent() ? "\u03A6" : node.parent.key));
        this.inOrderTreeWalk(node.right);
    }
    
    getLevel(node) {
        if (this.root === null || node === null) {
            return -1;
        }
        
        let iter = this.root;
        let level = 0;
        
        while (iter !== null && iter !== node) {
            if (node.key < iter.key) {
                iter = iter.left;
            }
            else {
                iter = iter.right;
            }
            
            level += 1;
        }
        
        return level;
    }
    
    get(val) {
        return this.search(this.root, val);
    }
    
    has(val) {
        return (this.search(this.root, val) !== null);
    }
};



